src/Microsoft.TestPlatform.CrossPlatEngine/Hosting/DefaultTestHostManager.cs

```diff
@@ -108,64 +68,67 @@ internal DefaultTestHostManager(Architecture architecture, Framework framework,
 
-        /// <inheritdoc/>
-        public void SetCustomLauncher(ITestHostLauncher customLauncher)
-        {
-            this.customTestHostLauncher = customLauncher;
-        }
-
-        /// <inheritdoc/>
-        public async Task<int> LaunchTestHostAsync(TestProcessStartInfo testHostStartInfo)
-        {
-            return await Task.Run(() => LaunchHost(testHostStartInfo), this.GetCancellationTokenSource().Token);
-        }
+        protected int ErrorLength { get; set; } = 1000;
 
-        /// <inheritdoc/>
-        private CancellationTokenSource GetCancellationTokenSource()
-        {
-            this.hostLaunchCTS = new CancellationTokenSource(TimeOut);
-            return this.hostLaunchCTS;
-        }
+        protected int TimeOut { get; set; } = 10000;
 
-        private int LaunchHost(TestProcessStartInfo testHostStartInfo)
+        /// <summary>
+        /// Callback on process exit
+        /// </summary>
+        private Action<Process, string> ErrorReceivedCallback => ((process, data) =>
         {
-            try
+            var exitCode = 0;
+            if (!string.IsNullOrEmpty(data))
             {
-                this.testHostProcessStdError = new StringBuilder(this.ErrorLength, this.ErrorLength);
-                EqtTrace.Verbose("Launching default test Host Process {0} with arguments {1}", testHostStartInfo.FileName, testHostStartInfo.Arguments);
-
-                if (this.customTestHostLauncher == null)
+                // if incoming data stream is huge empty entire testError stream, & limit data stream to MaxCapacity
+                if (data.Length > this.testHostProcessStdError.MaxCapacity)
                 {
-                    this.testHostProcess = this.processHelper.LaunchProcess(testHostStartInfo.FileName, testHostStartInfo.Arguments, testHostStartInfo.WorkingDirectory, testHostStartInfo.EnvironmentVariables, this.ErrorReceivedCallback);
+                    this.testHostProcessStdError.Clear();
+                    data = data.Substring(data.Length - this.testHostProcessStdError.MaxCapacity);
                 }
+
+                // remove only what is required, from beginning of error stream
                 else
                 {
-                    int processId = this.customTestHostLauncher.LaunchTestHost(testHostStartInfo);
-                    this.testHostProcess = Process.GetProcessById(processId);
+                    int required = data.Length + this.testHostProcessStdError.Length - this.testHostProcessStdError.MaxCapacity;
+                    if (required > 0)
+                    {
+                        this.testHostProcessStdError.Remove(0, required);
+                    }
                 }
 
+                this.testHostProcessStdError.Append(data);
             }
-            catch (OperationCanceledException ex)
+
+            if (this.processHelper.TryGetExitCode(process, out exitCode))
```

Comment:
Missing scenario: process doesn't write anything on `stderr` and exits -> HostExited should be called.